### ISSUE TITLE
[Question] Inconsistent training of Panda manipulation tasks

### ISSUE THREAD

**Issue by fikricanozgur** [DLR-RM/rl-baselines3-zoo#322]
### ❓ Question

Hello,

Training the Panda tasks (push, slide, pick&place) with TQC using the corresponding hyper-parameters in this repository sometimes does not converge to any meaningful behaviour and the success rate stays too low, see the plots below.  I wonder if this is normal and if there are other parameters that I can modify to make the trainings more consistent? 

Thanks!

![pp](https://user-images.githubusercontent.com/54752334/204493790-520e8cc8-24c7-4b60-a8ab-79c8cd82d7db.png)
![slide](https://user-images.githubusercontent.com/54752334/204494878-924f2909-30ab-4dca-805a-afa2e0da72e0.png)
![W B Chart 11_29_2022, 10_45_20 AM](https://user-images.githubusercontent.com/54752334/204495122-ee007dfc-c3fb-4762-9ba7-dc24233e210f.png)

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/rl-baselines3-zoo/issues) in the repo
- [X] I have read the [SB3 documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have read the [RL Zoo README](https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/README.md)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by fikricanozgur** [comment]
Hello, I would appreciate if you could give your comment on this question when you find the time. If you do not support this kind of inquires then kindly let me know and I would understand. Have a nice day.

**Comment by qgallouedec** [comment]
Hi, sorry for the delay, I've been busy these days. I'll answer your question as soon as possible.

**Comment by qgallouedec** [comment]
I am running several trainings to see if I can reproduce your curves. I want to make sure it's not an implementation error or a bug (both panda-gym and SB3). 
If it's clear on this side, then my guess is that for some runs, the agent is not exploring its environment enough. Maybe increasing the noise of the actions would give better results.

**Comment by qgallouedec** [comment]
PandaPush-v1 TQC: 
![PandaPush-v1_TQC](https://user-images.githubusercontent.com/45557362/206409012-b3b8e0b2-72ab-477f-96d9-d8daad761001.png)

I have run 6 trainings on PandaPush-v1 with TQC, and I did not get such variability in results. I advise you to double check that the hyperparameters of your two trainings are equal.

**Comment by fikricanozgur** [comment]
Alright, thanks for looking into it. I will check the hyperparameters again and see if I can reproduce your results.

**Comment by qgallouedec** [comment]
> using the corresponding hyper-parameters in this repository 

I also suspect that you did not use the hyperparameters advised in the zoo, because even your best execution on Push is well below the curves I got. (At 200k timesteps: about 60% versus 20%)

**Comment by fikricanozgur** [comment]
Yes, that is rather surprising to me as well. I am working on a project and using SB3 for trainings, it might be that I changed something in the code or in the hyperparameters and then forgot about it. I will repeat the trainings on a fresh installation and report the results back here.

**Comment by araffin** [comment]
> PandaPush-v1 TQC:

@qgallouedec could you share the runs and the command line you use? (so it's easier for @fikricanozgur to reproduce runs)
maybe would be good to move the runs to openrlbenchmark (ask @vwxyzjn  for access), we have a SB3 workspace here: https://wandb.ai/openrlbenchmark/sb3

I remember having some failure in the past (but not often), so I think it might depends on the random seed too.

**Comment by qgallouedec** [comment]
> @qgallouedec could you share the runs and the command line you use?

```shell
python train.py --algo tqc --env PandaPush-v1
```

(I also added `--track --wandb-project-name panda --n 500000` but it shouldn't affect the results.)

> maybe would be good to move the runs to openrlbenchmark (ask @vwxyzjn for access), we have a SB3 workspace here: https://wandb.ai/openrlbenchmark/sb3

I will :+1:

**Comment by qgallouedec** [comment]
Some updates: the 7th run seems to replicate your problem.

![W B Chart 12_8_2022, 1 18 44 PM](https://user-images.githubusercontent.com/45557362/206445084-54caabdc-2c2b-4f72-b408-a7ef5f908208.png)

For some seeds, it seems that we are reaching a kind of divergence situation. Here are the losses:

![Screenshot from 2022-12-08 13-20-01](https://user-images.githubusercontent.com/45557362/206445258-9dc14c8d-3882-4400-b454-fc075aaba78d.png)

**Comment by araffin** [comment]
> For some seeds, it seems that we are reaching a kind of divergence situation.

I've experienced that with SAC (and derivates) in the past, there are different solutions:
- one is to clip the actor output (gSDE solution) / l2 regularization on its output (was in the original implementation)
- another is to try to limit the standard deviation even more (or use `expln` solution with gSDE) to avoid explosion

**Comment by fikricanozgur** [comment]
L2 regularization on the weights (using AdamW) with gSDE seem to solve the divergent behavior observed previously. Thanks @araffin. 

![image](https://user-images.githubusercontent.com/54752334/206893915-28524d2b-383a-4a7e-8e18-b2dab62ad7bf.png)

**Comment by araffin** [comment]
Hello,
good to hear =)
could you share the hyperparameters used?
(my guess is that l2 regularization should be enough, gSDE might not be needed, especially as we are not using it as intended)

**Comment by fikricanozgur** [comment]
Hi,

I used the TQC algorithm with the default hyperparameters in the repo and turned gSDE on. I changed the optimizer to AdamW using its default weight decay of 0.01. Another change I made was to lower the distance_threshold of the Push task to 5mm which is 5cm normally.

**Comment by araffin** [comment]
@qgallouedec after some quick trials, I think we should change the default optimizer to `AdamW`.
For the figure below, I used:
```
python train.py --algo tqc --env PandaPush-v1 -P --seed 2
```
vs
```
python train.py --algo tqc --env PandaPush-v1 -P --seed 2 -params policy_kwargs:'dict(net_arch=[512, 512, 512], n_critics=2, optimizer_class=th.optim.AdamW)'
```
so the only difference is the optimizer, the seed for runs 2-3 and 4-5 are the same (seed=1915415480 and seed=2).
runs 2-4 is with Adam, runs 3-5 is with AdamW:
![Training_Success_Rate](https://user-images.githubusercontent.com/1973948/207403225-e47a4601-1f37-4cb6-8b15-54ceb2260cc4.png)


We would need more seeds to be sure (I couldn't reproduce the explosion using the W&B seed of your run though) but if the symptoms are loss explosion, then `AdamW` with weight decay should definitely be a good solution.

**Comment by qgallouedec** [comment]
Thanks for the suggestion. I’ll evaluate more precisely the impact of this new optimiser on the results (it is a good first use case of openrlbenchmark!).
If it is confirmed, we will change the optimiser.

**Comment by vwxyzjn** [comment]
Hey @qgallouedec fwiw, you can tag the experiments with version control information, then use openrlbenchmark to filter the experiments of the new tag. For example, you could tag the current sb3 experiments with `WANDB_TAGS=sb3-1.6.2`, then you can tag the experiments with the new optimiser with `pr-322`. Finally, you can do 
```
python -m openrlbenchmark.rlops \
    --filters '?we=openrlbenchmark&wpn=sb3&ceik=env&cen=algo&metric=rollout/ep_rew_mean' \
        'tqc?tag=sb3-1.6.2'
        'tqc?tag=pr-322'
    --env-ids HalfCheetahBulletEnv-v0 \
    --ncols 1 \
    --ncols-legend 2 \
    --output-filename compare.png \
    --report
```
We have a feature to tag experiments automatically here (https://github.com/vwxyzjn/cleanrl/blob/b558b2b48326d8bb8da7fa853914576ae4610f53/cleanrl_utils/benchmark.py#L38-L62)

See https://cleanrl-d978juk2k-vwxyzjn.vercel.app/advanced/rlops/ for more detail on the workflow.

**Comment by qgallouedec** [comment]
Great, thanks!!
I think it would be nice to add something similar in sb3 zoo.
